### PR TITLE
null safety

### PR DESCRIPTION
**Pull request by lig**
This is a work in progress.

`ordered_set` uses override to a branch in my fork until this is merged https://github.com/luanpotter/ordered_set/pull/5.

As for now, nothing is done in the flame code itself.

Todo:
- [x] prepare main `pubspec.yaml` file
- [x] migrate main dependencies
- [x] migrate main code
- [x] prepare examples `pubspec.yaml` files
- [x] migrate examples dependencies
- [x] migrate examples code
- [x] ensure CI passes
- [x] changelog entry
- [x] rebase on `master` with null safety backports merged in 
- [x] Dashbook null safety release

### CODE REVIEW

**Comment by lig** [comment]
Depends on https://github.com/luanpotter/ordered_set/pull/5

**Comment by unger1984** [comment]
I fork your branch and complete null-safety prepare. Please, merge my work to your branch https://github.com/unger1984/flame/tree/null-safety

File: lib/assets/assets_cache.dart [hunk 1]
```diff
@@ -54,22 +54,27 @@ class AssetsCache {
...
 }
 
-class _StringAsset extends _Asset<String> {}
+class _StringAsset extends _Asset<String> {
+  _StringAsset(String value) : super(value);
+}
 
-class _BinaryAsset extends _Asset<List<int>> {}
+class _BinaryAsset extends _Asset<List<int>> {
+  _BinaryAsset(List<int> value) : super(value);
```
Comment (luanpotter) [inline lib/assets/assets_cache.dart:79]: bits like this can be extracted to a separate PR to master, to reduce the null safety diff. also any API changes we might wanna do to better support null-safety

Comment (lig) [reply lib/assets/assets_cache.dart:79]: :+1: reducing null safety patchset size is a great idea as it is going to hang being separated for a while.

I'd be happy to target some changes to `master`. Just point me out to places like that. I've got you on this one.

File: lib/effects/effects.dart [hunk 2]
```diff
@@ -29,9 +29,14 @@ abstract class ComponentEffect<T extends Component> {
   final bool isRelative;
   final bool _initialIsInfinite;
   final bool _initialIsAlternating;
-  double percentage;
-  double curveProgress;
-  double peakTime;
+  double? percentage;
+
+  /// FIXME: is it a safe default?
+  double curveProgress = 0.0;
```
Comment (lig) [inline lib/effects/effects.dart:35]: @luanpotter is it a safe default?

File: lib/effects/effects.dart [hunk 3]
```diff
@@ -29,9 +29,14 @@ abstract class ComponentEffect<T extends Component> {
...
+
+  /// FIXME: Where it's been set? Let's try defaulting it to 0.0 for now as at least
+  /// some subclasses set it in `initialize` method
+  double peakTime = 0.0;
```
Comment (lig) [inline lib/effects/effects.dart:39]: @luanpotter does it look ok to you to use a default value here?

File: lib/particles/computed_particle.dart [hunk 4]
```diff
@@ -17,9 +15,10 @@ class ComputedParticle extends Particle {
   ParticleRenderDelegate renderer;
 
   ComputedParticle({
-    @required this.renderer,
-    double lifespan,
-    Duration duration,
+    required this.renderer,
+    double? lifespan,
+    /// FIXME: why `duration` is here?
+    Duration? duration,
```
Comment (lig) [inline lib/particles/computed_particle.dart:21]: @luanpotter I didn't get why `duration` is here at all

**Review by lig — COMMENTED** [review]
@luanpotter could you please answer the questions?

**Comment by lig** [comment]
@luanpotter should I target #553 to the `develop` branch?

File: lib/effects/effects.dart (hunk 2 above)
Comment (spydon) [reply lib/effects/effects.dart:35]: Yes, it should be fine since.

File: lib/effects/effects.dart (hunk 3 above)
Comment (spydon) [reply lib/effects/effects.dart:39]: It should be fine here too since it is initialized before the first `update` is called.

**Comment by lig** [comment]
So, it turns out that `doc/examples/widgets` has `dashbook` as its dependency. Need to migrate this one.

**Comment by lig** [comment]
`dashbook` null safety PR: https://github.com/erickzanardo/dashbook/pull/25

File: lib/assets/assets_cache.dart (hunk 1 above)
Comment (lig) [reply lib/assets/assets_cache.dart:79]: #553 has been merged. I've rebased the PR. This doesn't show up in diff as the `flame-engine:upstream` branch wasn't rebased on `develop`.

**Comment by lig** [comment]
Until `dashbook` has null safety support I'm marking this as a Draft.

File: README-RU.md [hunk 5]
```diff
@@ -0,0 +1,113 @@
+[![Pub](https://img.shields.io/pub/v/flame.svg?style=popout)](https://pub.dartlang.org/packages/flame) ![Test](https://github.com/flame-engine/flame/workflows/Test/badge.svg?branch=master&event=push) [![Discord](https://img.shields.io/discord/509714518008528896.svg)](https://discord.gg/pxrBmy4)
```
Comment (luanpotter) [inline README-RU.md:1]: we should extract this to a separated PR @lig

File: doc/examples/effects/infinite_effects/lib/main.dart [hunk 6]
```diff
@@ -18,9 +18,9 @@ void main() async {
 }
 
 class MyGame extends BaseGame with TapDetector {
-  Square greenSquare;
-  Square redSquare;
-  Square orangeSquare;
+  Square? greenSquare;
+  Square? redSquare;
+  Square? orangeSquare;
```
Comment (luanpotter) [inline doc/examples/effects/infinite_effects/lib/main.dart:23]: should these all be late? or even move the initializer to here

File: doc/examples/gestures/lib/main_multitap_advanced.dart [hunk 7]
```diff
@@ -70,12 +72,14 @@ class MyGame extends Game with MultiTouchTapDetector, MultiTouchDragDetector {
...
   void onPanEnd(DragEndDetails details) {
-    _panRect = Rect.fromLTRB(
-      _start.dx,
-      _start.dy,
-      _end.dx,
-      _end.dy,
-    );
+    if (_start != null && _end != null) {
+      _panRect = Rect.fromLTRB(
+        _start!.dx,
```
Comment (luanpotter) [inline doc/examples/gestures/lib/main_multitap_advanced.dart:77]: do we need ! if we already asserted _start != null?

File: lib/text_config.dart [hunk 8]
```diff
@@ -88,9 +88,12 @@ class TextConfig {
     Vector2 p, {
     Anchor anchor = Anchor.topLeft,
   }) {
-    final material.TextPainter tp = toTextPainter(text);
-    final Vector2 translatedPosition = anchor.translate(p, tp.size.toVector2());
-    tp.paint(canvas, translatedPosition.toOffset());
+    final material.TextPainter? tp = toTextPainter(text);
```
Comment (luanpotter) [inline lib/text_config.dart:91]: doesn't toTextPainter returns non-null?

File: lib/particles/moving_particle.dart [hunk 9]
```diff
@@ -30,8 +29,10 @@ class MovingParticle extends CurvedParticle with SingleChildParticle {
   @override
   void render(Canvas c) {
     c.save();
-    final Offset current = Offset.lerp(from, to, progress);
-    c.translate(current.dx, current.dy);
+    final Offset? current = Offset.lerp(from, to, progress);
```
Comment (luanpotter) [inline lib/particles/moving_particle.dart:32]: how could this return null?

File: lib/particles/computed_particle.dart (hunk 4 above)
Comment (luanpotter) [reply lib/particles/computed_particle.dart:22]: cc @spydon

Comment (luanpotter) [reply lib/particles/computed_particle.dart:22]: sorry I confused particles with effects haha. I will take a look, maybe @av knows?

File: README-RU.md (hunk 5 above)
Comment (lig) [reply README-RU.md:1]: Oh boy. I didn't notice. Agree. @unger1984 can you do this?

File: doc/examples/effects/infinite_effects/lib/main.dart (hunk 6 above)
Comment (lig) [reply doc/examples/effects/infinite_effects/lib/main.dart:23]:
```dart
  MyGame()
      : greenSquare =
            Square(Paint()..color = const Color(0xAA338833), Vector2.all(100)),
        redSquare =
            Square(Paint()..color = const Color(0xAA883333), Vector2.all(200)),
        orangeSquare = Square(
            Paint()..color = const Color(0xAABB6633), Vector2(200, 400)) {
    add(greenSquare);
    add(redSquare);
    add(orangeSquare);
  }
```

I'd rather do `late` than that.

File: doc/examples/gestures/lib/main_multitap_advanced.dart (hunk 7 above)
Comment (lig) [reply doc/examples/gestures/lib/main_multitap_advanced.dart:77]: well, `flutter analyze` fails to see that. it happens in some cases. I still didn't grasp when it does that completely but in this case, we need to be explicit.

File: lib/text_config.dart (hunk 8 above)
Comment (lig) [reply lib/text_config.dart:91]: You're right. I looked at the check after that and put that there. Removing nullable and the check.

File: lib/particles/moving_particle.dart (hunk 9 above)
Comment (lig) [reply lib/particles/moving_particle.dart:32]: This one is interesting. If both `from` and `to` are `null` it returns `null`. However. we use non-null values for both. So, in our case, it never returns null.

Forcing non-null here then.

File: README-RU.md (hunk 5 above)
Comment (unger1984) [reply README-RU.md:1]: It`s was merged in PR #551

Comment (lig) [reply README-RU.md:1]: > It`s was merged in PR #551

I see. It could be that I've rebased my branch on `develop`.

Comment (lig) [reply README-RU.md:1]: @luanpotter could you please move the `null-safety` branch forward with `develop`?

Comment (spydon) [reply README-RU.md:1]: @lig updated it now, there will most likely be several branches merged to develop today though.
If you ping me tomorrow I can update it again if I forget.

Comment (lig) [reply README-RU.md:1]: @spydon wilco, thanks! After this PR merged `null-safety` will be diverted and each rebase is going to be a separate PR, I guess.

**Comment by spydon** [comment]
@lig Updated the underlying branch to be based on `master` since we scrapped `develop`.

**Comment by lig** [comment]
> @lig Updated the underlying branch to be based on `master` since we scrapped `develop`.

Thanks for letting me know. I'll rebase then.

**Comment by lig** [comment]
@luanpotter it's ready

**Comment by spydon** [comment]
@lig it looks like you have made a lot of progress!
I just updated the underlying branch with master.

**Comment by lig** [comment]
@spydon I'm working on fixes after the branch update and it looks like Flame could benefit from being null safe first.

There are places where null safety catches some possible errors.

For instance, here https://github.com/flame-engine/flame/blob/null-safety/lib/components/parallax_component.dart#L104 `_imageSize` could be `null` in case `future` resolves after `resize` and before `update`. `resize` won't set `_imageSize` but `update` will try to use it. This particular case could be avoided using an additional check in `update`.

I will update the entire PR and add comments on such places lately.

File: lib/components/parallax_component.dart [hunk 10]
```diff
@@ -51,7 +51,11 @@ class ParallaxLayer {
 
   bool loaded() => _image != null;
 
-  Vector2 currentOffset() => _scroll;
+  /// FIXME: added this as there are a lot of things being set in resize first time that
+  /// are being used in update
+  bool sized = false;
```
Comment (lig) [inline lib/components/parallax_component.dart:56]: Added this. Otherwise, `update` would be bloated with null checks for `_scale`, `_imagesSize`, `_paintArea`

**Comment by lig** [comment]
@spydon I've updated the code.

See two open questions in the review.

Tests pass locally on my machine on the same flutter/dart version. It looks like something is off a bit. Maybe it doesn't run fast enough in Github action?

**Comment by spydon** [comment]
> @spydon I've updated the code.
> 
> See two open questions in the review.
> 
> Tests pass locally on my machine on the same flutter/dart version. It looks like something is off a bit. Maybe it doesn't run fast enough in Github action?

The tests worked when I re-ran them, but they shouldn't have failed in the first place so we should be careful when reviewing the changes of the effects code.

**Comment by spydon** [comment]
> @spydon I'm working on fixes after the branch update and it looks like Flame could benefit from being null safe first.
> 
> There are places where null safety catches some possible errors.
> 
> For instance, here https://github.com/flame-engine/flame/blob/null-safety/lib/components/parallax_component.dart#L104 `_imageSize` could be `null` in case `future` resolves after `resize` and before `update`. `resize` won't set `_imageSize` but `update` will try to use it. This particular case could be avoided using an additional check in `update`.
> 
> I will update the entire PR and add comments on such places lately.

These cases would be good to do separate PRs for since they could be merged to master already.

**Comment by lig** [comment]
> > @spydon I'm working on fixes after the branch update and it looks like Flame could benefit from being null safe first.
> > There are places where null safety catches some possible errors.
> > For instance, here https://github.com/flame-engine/flame/blob/null-safety/lib/components/parallax_component.dart#L104 `_imageSize` could be `null` in case `future` resolves after `resize` and before `update`. `resize` won't set `_imageSize` but `update` will try to use it. This particular case could be avoided using an additional check in `update`.
> > I will update the entire PR and add comments on such places lately.
> 
> These cases would be good to do separate PRs for since they could be merged to master already.

wilco

**Comment by lig** [comment]
Rebased onto the latest `master`

**Comment by lig** [comment]
Tests are off again. It looks like some fuzzy logic is too strict or the test should have more control.

For the record
```
00:16 +41: /home/runner/work/flame/flame/test/effects/sequence_effect_test.dart: SequenceEffect can be infinite                                                                                        
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
The following TestFailure object was thrown running a test:
  Expected: a numeric value within <0.01> of <0.0>
  Actual: <0.02412071442309163>
   Which:  differs by <0.02412071442309163>
Size is not correct

When the exception was thrown, this was the stack:
#4      effectTest (file:///home/runner/work/flame/flame/test/effects/effect_test_utils.dart:60:5)
<asynchronous suspension>
...

This was caught by the test expectation on the following line:
  file:///home/runner/work/flame/flame/test/effects/effect_test_utils.dart line 60
The test description was:
  SequenceEffect can be infinite
════════════════════════════════════════════════════════════════════════════════════════════════════

00:16 +41 -1: /home/runner/work/flame/flame/test/effects/sequence_effect_test.dart: SequenceEffect can be infinite [E]                                                                                 
  Test failed. See exception logs above.
  The test description was: SequenceEffect can be infinite
```

**Comment by spydon** [comment]
> Tests are off again. It looks like some fuzzy logic is too strict or the test should have more control.
> 
> Could be a potential bug in the null safety code only as well.
> 
> For the record
> 
> ```
> 00:16 +41: /home/runner/work/flame/flame/test/effects/sequence_effect_test.dart: SequenceEffect can be infinite                                                                                        
> ══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
> The following TestFailure object was thrown running a test:
>   Expected: a numeric value within <0.01> of <0.0>
>   Actual: <0.02412071442309163>
>    Which:  differs by <0.02412071442309163>
> Size is not correct
> 
> When the exception was thrown, this was the stack:
> #4      effectTest (file:///home/runner/work/flame/flame/test/effects/effect_test_utils.dart:60:5)
> <asynchronous suspension>
> ...
> 
> This was caught by the test expectation on the following line:
>   file:///home/runner/work/flame/flame/test/effects/effect_test_utils.dart line 60
> The test description was:
>   SequenceEffect can be infinite
> ════════════════════════════════════════════════════════════════════════════════════════════════════
> 
> 00:16 +41 -1: /home/runner/work/flame/flame/test/effects/sequence_effect_test.dart: SequenceEffect can be infinite [E]                                                                                 
>   Test failed. See exception logs above.
>   The test description was: SequenceEffect can be infinite
> ```

Something must be affecting the tests on this branch, because they are not flaky on master.

**Comment by lig** [comment]
> Something must be affecting the tests on this branch, because they are not flaky on master.

From the latest build in `master`
https://github.com/flame-engine/flame/runs/1637361349

```
00:11 +39: /home/runner/work/flame/flame/test/effects/sequence_effect_test.dart: SequenceEffect can alternate and be infinite                                                                          
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
The following TestFailure object was thrown running a test:
  Expected: a numeric value within <0.01> of <5.0>
  Actual: <4.984999999999999>
   Which:  differs by <0.015000000000000568>
Angle is not correct

When the exception was thrown, this was the stack:
#4      effectTest (file:///home/runner/work/flame/flame/test/effects/effect_test_utils.dart:55:5)
<asynchronous suspension>
...

This was caught by the test expectation on the following line:
  file:///home/runner/work/flame/flame/test/effects/effect_test_utils.dart line 55
The test description was:
  SequenceEffect can alternate and be infinite
════════════════════════════════════════════════════════════════════════════════════════════════════

00:11 +39 -1: /home/runner/work/flame/flame/test/effects/sequence_effect_test.dart: SequenceEffect can alternate and be infinite [E]                                                                   
  Test failed. See exception logs above.
  The test description was: SequenceEffect can alternate and be infinite
```

**Comment by spydon** [comment]
> > Something must be affecting the tests on this branch, because they are not flaky on master.
> 
> From the latest build in `master`
> https://github.com/flame-engine/flame/runs/1637361349
> 
> ```
> 00:11 +39: /home/runner/work/flame/flame/test/effects/sequence_effect_test.dart: SequenceEffect can alternate and be infinite                                                                          
> ══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
> The following TestFailure object was thrown running a test:
>   Expected: a numeric value within <0.01> of <5.0>
>   Actual: <4.984999999999999>
>    Which:  differs by <0.015000000000000568>
> Angle is not correct
> 
> When the exception was thrown, this was the stack:
> #4      effectTest (file:///home/runner/work/flame/flame/test/effects/effect_test_utils.dart:55:5)
> <asynchronous suspension>
> ...
> 
> This was caught by the test expectation on the following line:
>   file:///home/runner/work/flame/flame/test/effects/effect_test_utils.dart line 55
> The test description was:
>   SequenceEffect can alternate and be infinite
> ════════════════════════════════════════════════════════════════════════════════════════════════════
> 
> 00:11 +39 -1: /home/runner/work/flame/flame/test/effects/sequence_effect_test.dart: SequenceEffect can alternate and be infinite [E]                                                                   
>   Test failed. See exception logs above.
>   The test description was: SequenceEffect can alternate and be infinite
> ```

Ooh, hadn't seen that, I'll have a look at that test tomorrow if I get time. It is always the same one right?

**Comment by lig** [comment]
> Ooh, hadn't seen that, I'll have a look at that test tomorrow if I get time. It is always the same one right?

As far as I'm aware it is always the same one.

File: lib/particles/computed_particle.dart (hunk 4 above)
Comment (lig) [reply lib/particles/computed_particle.dart]: Removed for good

**Comment by lig** [comment]
Removed dependency override for Dashbook

**Comment by lig** [comment]
@luanpotter @spydon @erickzanardo This one is ready.
PR says the branch has 19 commits. This is because the target branch is outdated. The PR is rebased on the latest `master` branch and effectively contains a single commit.

**Comment by spydon** [comment]
> @luanpotter @spydon @erickzanardo This one is ready.
> PR says the branch has 19 commits. This is because the target branch is outdated. The PR is rebased on the latest `master` branch and effectively contains a single commit.

Updated the null-safety branch to latest master

**Comment by lig** [comment]
`+564 −592`
less code -- fewer bugs:)

**Comment by lig** [comment]
Updated to rc6.

Going to extract `lib/src/components/mixins/hitbox.dart` fixes into a separate PR targeted to `master` (cc: @spydon).

**Comment by spydon** [comment]
> Updated to rc6.
> 
> Going to extract `lib/src/components/mixins/hitbox.dart` fixes into a separate PR targeted to `master` (cc: @spydon).
> 

Might not be worth doing that already, because there will be a lot of changes to that in the collision detection branch.

**Comment by lig** [comment]
> Might not be worth doing that already, because there will be a lot of changes to that in the collision detection branch.

It's not a lot but I would like to see that early in to avoid spreading possible nullable values in null unsafe code.

**Comment by lig** [comment]
🎉
